Port to Kotlin Multiplatform

Ercalvez opened the pull request:
Work still need to be done:
- [ ] Publishing and signing setup

`library-android` has been moved to `androidMain`
`FileSystem.SYSTEM` is not available in commonMain.

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/Feature.kt:
```diff
@@ -0,0 +1,24 @@
+package de.westnordost.osmfeatures
+
+interface Feature {
```
```suggestion
/** Subset of a feature as defined in the iD editor
 *  https://github.com/ideditor/schema-builder#preset-schema
 *  with only the fields helpful for the dictionary */
interface Feature {
```

Add comment

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/BaseFeature.kt:
```diff
@@ -0,0 +1,44 @@
+
+        init {
+                this.canonicalTerms = terms.map { term -> StringUtils.canonicalize(term)}
+        }
```
```suggestion
        final override val canonicalTerms: List<String> = terms.map { term -> StringUtils.canonicalize(term)}
```

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/Feature.kt:
```diff
@@ -0,0 +1,24 @@
+    val addTags: Map<String, String>
+    val removeTags: Map<String, String>
+    val canonicalNames: List<String>
+    val canonicalTerms: List<String>?
```
```suggestion
    val canonicalTerms: List<String>
```
Any reason why this is nullable? If not, I prefer if this was not nullable.

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/LocalizedFeature.kt:
```diff
@@ -0,0 +1,59 @@
+    override val names: List<String>,
+    override val terms: List<String>
+) :
+    Feature {
```
```suggestion
) : Feature {
```

Format

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/LocalizedFeature.kt:
```diff
@@ -0,0 +1,59 @@
+            canonicalTerms.add(StringUtils.canonicalize(term))
+        }
+        this.canonicalTerms = canonicalTerms.toList()
+    }
```
```suggestion
    override val canonicalNames: List<String> = names.map { name -> StringUtils.canonicalize(name) }
    override val canonicalTerms: List<String> = terms.map { term -> StringUtils.canonicalize(term) }
```

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/StartsWithStringTree.kt:
```diff
@@ -0,0 +1,105 @@
+ * [ "foobar", "foo", ...]
+ * u ->
+ * [ "fu", "funicular", ... ]
+</pre> */
```
```suggestion
 * <pre>
 * f ->
 *   [ "f" ]
 *   o ->
 *     o ->
 *       [ "foobar", "foo", ...]
 *   u ->
 *     [ "fu", "funicular", ... ]
 * </pre> */
```
format

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/StringUtils.kt:
```diff
@@ -0,0 +1,23 @@
+
+
+
+
```
```suggestion
```

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/ContainedMapTree.kt:
```diff
@@ -0,0 +1,156 @@
+ * supermarket ->
+ * #5
+ * ...
+</pre> *
```
```suggestion
 * [
 *   #1 (amenity -> bicycle_parking),
 *   #2 (amenity -> bicycle_parking, bicycle_parking -> shed),
 *   #3 (amenity -> bicycle_parking, bicycle_parking -> lockers),
 *   #4 (amenity -> taxi),
 *   #5 (shop -> supermarket),
 * ]
 * </pre>
 * ...the tree internally looks like this:
 * <pre>
 * amenity ->
 *   bicycle_parking ->
 *     #1
 *     bicycle_parking ->
 *       shed ->
 *         #2
 *       lockers ->
 *         #3
 *   taxi ->
 *     #4
 * shop ->
 *   supermarket ->
 *     #5
 * ...
* </pre>
```
format

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/BaseFeature.kt:
```diff
@@ -0,0 +1,44 @@
+        final override var geometry: List<GeometryType>,
+        private val _icon: String? = "",
+        private val _imageURL: String? = "",
+        private val _names: List<String>,
```
Huh, why? I don't see this in the original Java implementation.

westnordost reviewed (commented):
A very incomplete review while in the train. I think when home I'll rather check out the branch and correct all the small issues I find myself and only note the bigger ones (if any) here on github, more efficient that way.

It's a good idea to have the Android-specific stuff in the Android-part of the library itself, makes sense.

Four general comments:

1. The directory name looks suspect: `library/src/commonMain/kotlin/de.westnordost.osmfeatures/...` - i.e. why are there dots in the name instead of subdirectories? It's the first time that I see this. I checked [the Kotlin coding conventions](https://kotlinlang.org/docs/coding-conventions.html#source-code-organization) and didn't see that mentioned.

2. Regarding the build.gradle.kts: I recently created another kotlin multiplatform library, maybe you could copy it from there: https://github.com/westnordost/osm-opening-hours

3. <strike>Some fields, parameters and variables you made nullable even though this looks like an oversight. In case it is not, comments should be added to the source to explain / document it.</strike>

4. <strike>With kotlinx-serialization-json it is not necessary (nor recommended, as far as I know) to do JSON parsing by hand. You can simply annotate the data model how it should be (de)serialized. See https://kotlinlang.org/docs/serialization.html#serialize-and-deserialize-json . Did you not know this, or is there some reason why you chose to parse it by hand?</strike>

westnordost commented on library/src/commonMain/kotlin/de.westnordost.osmfeatures/Locale.kt:
```diff
@@ -0,0 +1,17 @@
+    val script: String? = null
+) {
+        companion object {
+            val default: Locale? = null
```
`Locale.default` (or any other API) should return the user's default locale!

So, this must either be solved using `expect` / `actual` to create the default locale on the different platforms, or remove any functionality that is associated with working with the default locale. (Didn't check yet what this would encompass)

Ercalvez replied:
I'll look into it

westnordost replied:
See point 4 of my review, I already implemented it if we use String instead of an own Locale data class. And I tend towards that solution, because it is most portable and also forward compatible (e.g. in case iD preset localizations will appear for other weird language/sublanguage combinations that the IETF language tag allows

westnordost replied:
I did merge that now into this branch.

westnordost commented on library/src/androidMain/kotlin/de/westnordost/osmfeatures/FileSystemAccess.kt:
```diff
@@ -0,0 +1,7 @@
+import okio.FileSystem
+
+actual fun fileSystem(): FileSystem {
+    return FileSystem.SYSTEM
```
What's the use of this if it is the same on every platform?

Ercalvez replied:
`FileSystem.SYSTEM` is not available in commonMain

westnordost reviewed (changes requested):
Alright, I finished a thorough review. Since I corrected small issues myself, I also took the liberty to do more, which means that not everything I changed was objectively an issue, in fact, most of it was just about my preference on how the code should be formatted.
I reviewed all code except the `*JsonParser` classes and associated tests because as noted earlier, these could maybe be completely removed. Of the build.gradle stuff, I reviewed nothing yet.

- `synchronized` is only supported on JVM. So, `synchronizedGetOrCreate` in `CollectionUtils` won't work. In Kotlin, there's instead `Lazy` that should be used in places where there is a lazily initialized field.

- Most files are missing all those imports, such as even `import kotlin.collections.List`. What about that? Is the possibility to omit imports from the standard library some Kotlin feature I don't know about?

- `LivePresetDataAccessAdapter`, `IDPresetsJsonParserTest.kt` and `IDPresetsTranslationJsonParserTest.kt` could go into `commonTest`. The only thing that is missing for that is a multiplatform http client as a test dependency. You could use Ktor Client for that. (For now I moved all tests that use java's `URL` into `jvmTest`

- I think that `Locale` could just as well be a string, specifically a [IETF language tag](https://en.wikipedia.org/wiki/IETF_language_tag), like "fr-CA" . What do you think?

Ercalvez commented:
> Alright, I finished a thorough review. Since I corrected small issues myself, I also took the liberty to do more, which means that not everything I changed was objectively an issue, in fact, most of it was just about my preference on how the code should be formatted.
> 
> I reviewed all code except the `*JsonParser` classes and associated tests because as noted earlier, these could maybe be completely removed. Of the build.gradle stuff, I reviewed nothing yet.
> 
> 
> 
> - `synchronized` is only supported on JVM. So, `synchronizedGetOrCreate` in `CollectionUtils` won't work. In Kotlin, there's instead `Lazy` that should be used in places where there is a lazily initialized field.
> 
> 
> 
> - Most files are missing all those imports, such as even `import kotlin.collections.List`. What about that? Is the possibility to omit imports from the standard library some Kotlin feature I don't know about?
> 
> 
> 
> - `LivePresetDataAccessAdapter`, `IDPresetsJsonParserTest.kt` and `IDPresetsTranslationJsonParserTest.kt` could go into `commonTest`. The only thing that is missing for that is a multiplatform http client as a test dependency. You could use Ktor Client for that. (For now I moved all tests that use java's `URL` into `jvmTest`
> 
> 
> 
> - I think that `Locale` could just as well be a string, specifically a [IETF language tag](https://en.wikipedia.org/wiki/IETF_language_tag), like e.g. "fr-CA" . What do you think? (I already did the necessary changes in the kmp-language-tag branch)
- Yes `kotlin.collections` is imported by default in every Kotlin file: [Default imports](https://kotlinlang.org/docs/packages.html#default-imports)
- Regarding `Locale`, I also think it would be a good idea to rely on a language tag only
- I am going to modify the tests to use Ktor client instead of `java.URL`

westnordost commented:
> synchronized is only supported on JVM. So, synchronizedGetOrCreate in CollectionUtils won't work. In Kotlin, there's instead Lazy that should be used in places where there is a lazily initialized field.

Hmm, maybe Lazy is not the right thing... Kotlin's reply to "where multiplatform synchronized??" seems to be "use coroutine with mutex".... but, hmm, then the complete API would be suspending...

So, uh... maybe try if it works with Lazy after all.

westnordost commented:
> Yes kotlin.collections is imported by default in every Kotlin file: [Default imports](https://kotlinlang.org/docs/packages.html#default-imports)

Wow, that's cool, I didn't know that! Thank you for the link, let's see if I missed anything else.

Ercalvez commented:
I pushed new commits. Now tests use ktor to retrieve json files online (these few tests have been moved to `commonTest`). This test fails:
```kotlin
@Test
    fun load_features_only_defaults() {
        val features: List<BaseFeature> = parse("one_preset_min.json")
        assertEquals(1, features.size)
        val feature: Feature = features[0]
        assertEquals("some/id", feature.id)
        assertEquals(mapOf("a" to "b", "c" to "d"), feature.tags)
        assertEquals(listOf(GeometryType.POINT), feature.geometry)
        assertTrue(feature.includeCountryCodes.isEmpty())
        assertTrue(feature.excludeCountryCodes.isEmpty())
        assertEquals("", feature.name)
        assertEquals("", feature.icon)
        assertEquals("", feature.imageURL)
```
because `feature.name` is `null` instead of being an empty string. Is this the expected behaviour?

westnordost commented:
Hmm, I did change the API a bit (`icon` and `imageURL` can be null), but I think it makes sense that `name` can never be null. So, I adapted the parser.

Ercalvez commented:
I pushed a new version which uses the `lazy` operator for the synchronized map

westnordost commented:
Oh, wow. That works? Interesting! Well, that's certainly not how lazy is intended to be used. Hmm... actually, I think that doesn't work. As in, that is not thread-safe. Yes, the default thread safety mode of `lazy()` is `LazyThreadSafetyMode.SYNCHRONIZED` but this of course counts for every instance of lazy individually. Since you initialize a new instance on each function call, the lazy is not only never initialized and thus always calls the `createFn`, it is also not synchronized, i.e. it should have the same result as when you just leave out the `val value by lazy { ... }` block. But the point of having it lazy is to synchronize the creation.

Why do we want to synchronize the creation? Because the loading of a new language, a new country presets file is an IO operation that is potentially costly, so we want to only ever execute it once, not multiple times in case of parallelism.

The way to solve it is probably making the fields that use this extension function to actually be maps with a lazy value, i.e `Map<Key, Lazy<Value>>`.

Ercalvez commented:
I pushed commits to rearrange folders (`de.westnordost.osmfeatures` -> `de/westnordost/osmfeatures`). 
For `synchronizedGetOrCreate`, I think this implementation shoud be thread-safe:
```kotlin
internal fun <K, V> MutableMap<K, V>.synchronizedGetOrCreate(key: K, createFn: (K) -> V): V {
     val value = get(key)
      if (value == null) {
          val answer = lazy {
              if (get(key) == null) {
                  put(key, createFn(key))
              }
              get(key)
          }
          return answer.value as V
      }
      else {
          return value
      }
 }
```

westnordost commented:
No, sorry, it is not thread-safe. Two threads can enter line 4, both then create a new answer lazy (each is a new instance) and then both execute the createFn and put their value into the map.

westnordost commented:
So, this is what I have in mind.

```kotlin
internal fun <K, V> MutableMap<K, Lazy<V>>.getOrPutLazy(key: K, createFn: (K) -> V): V {
    if (!containsKey(key)) {
        put(key, lazy { createFn(key) })
    }
    return get(key)!!.value
}


// or actually, we can just use e.g.
val myLazyMap: MutableMap<String, Lazy<Things>>
//...
val myLazyThingsForKey = myLazyMap.getOrPut(key) { lazy { createTheThings() }}
// ... i.e. do not need an own extension function
```

This is of course still not entirely threadsafe as e.g. two threads could pass line 2, then thread one reaches line 5 and thus executes `createFn`, then thread two executes line 3, which would lead to another execution of `createFn`.
But the idea is that the lazy is constructed and added to the map before it is used, or in other words, that the time it takes to add the lazy to the map is as short as possible. I.e. in your suggestion, the value is only added to the map after the expensive calculation. This means, while the calculation is made, any thread that enters this function will also find that the value is not initialized yet and also initialize it in parallel. Here, the lazy is added immediately. If the lazy was in the map from the start and/or way before the lazy is executed, it would be even more safe, of course.
I am not a fan of non-100% solutions, but the above would be okay to me. 

A real threadsafe solution, I guess, could only be achieved by using https://github.com/Kotlin/kotlinx-atomicfu 
They write, that using this does not actually add any runtime dependency, which is nice, but it looks like this needs a bit of setup because it is not a simple library but a compiler plugin. Also, the `ReentrantLock` / `SynchronizedObject` from that library is marked as experimental (API might be changed significantly or removed), which I usually avoid, but if you prefer using atomicfu for a real threadsafe solution, that would be okay by me, too.

Ercalvez commented:
`lazy { createFn }` is still a good solution. As you said, it is not entirely thread safe, but since it is put immediately in the `Map`, it would rarely occur. It is preferable over introducing an experimental API

westnordost commented:
Yes, agree, although I may be forced to use atomicfu even if it is still experimental in StreetComplete, because in the data layer, we use `synchronized` quite a bit. I am hoping to hold off on this until it is classified as at least alpha... :-/

Ercalvez commented:
Unit tests failed for non-localized Features. `find_entry_by_tags` for example. I'm trying to find out why.

westnordost commented:
I didn't notice, before. Maybe this has something to do with your last change(s)?

westnordost commented:
If you need help, or you would like someone else to finish this PR, please say so. Maybe others are interested in contributing here if called onto it, or otherwise I can finish it myself.

Your work so far has been valuable and it is much appreciated! At the same time, I want to move this forward so that I can dedicate myself towards the next steps in creating an iOS port for the app.

Ercalvez commented:
I investigated on why some unit tests failed but I haven't found any reason why :(

There is the implementation of `atomicfu` to do, but maybe you'll be more suited to do that I guess.

westnordost commented:
Ok, then I'll finish this some time soon, maybe in two weeks. (Lazy is fine, though)

westnordost commented:
Alright, merged. In the end, the line of code count is only 72% of what we had for Java. 🎉

Since we made breaking changes to the API anyway (Locale -> String), I also changed a few more things that were bothering me.

For the time being, the initialization of the multiplatform version is about 15% slower than before. I think this is mainly due to whole file read into a string first, then into a DOM and then only in the final data structure. At least the first step should be solvable once kotlinx-serialization-json supports directly reading `kotlinx.io.Source`.

Ercalvez commented:
> Alright, merged. In the end, the line of code count is only 72% of what we had for Java. 🎉
> 
> 
> 
> Since we made breaking changes to the API anyway (Locale -> String), I also changed a few more things that were bothering me.
> 
> 
> 
> For the time being, the initialization of the multiplatform version is about 15% slower than before. I think this is mainly due to whole file read into a string first, then into a DOM and then only in the final data structure. At least the first step should be solvable once kotlinx-serialization-json supports directly reading `kotlinx.io.Source`.

Wow, 72% reduction is impressive 🎉

westnordost commented:
Nono, reduction **TO** 72%
